### PR TITLE
Delay Deletion of a Pod until volumes are cleaned up

### PR DESCRIPTION
**Pull request by dashpole**
#41436 fixed the bug that caused #41095 and #40239 to have to be reverted.  Now that the bug is fixed, this shouldn't cause problems.

 @vishh @derekwaynecarr @sjenning @jingxu97 @kubernetes/sig-storage-misc

### CODE REVIEW

**Comment by jingxu97** [comment]
@dashpole Could you put some details about this issue you discovered in the previous PR so that we could learn something from it? Thanks!

**Comment by dashpole** [comment]
@jingxu97  The issue from the previous PR, as shown [here](https://github.com/kubernetes/kubernetes/pull/40239#issuecomment-279522602), was that some updates to the API server were breaking the invariant that pod status always progresses.  For example, Status can go PENDING -> RUNNING -> FAILED, but should never go PENDING -> RUNNING -> FAILED -> RUNNING, which was what was happening in the previous PR.  By adding debugging statements, and reproducing the bug, I was able to see that calls to statusManager.SetPodStatus, was correctly updating the status to FAILED, and that a call to statusManager.TerminatePod was resetting the status to RUNNING.  This led to the discovery of #41436 in TerminatePod, which I believe has solved the problem.

File: pkg/kubelet/status/status_manager.go [hunk 1]
```diff
@@ -468,6 +472,15 @@ func (m *manager) needsUpdate(uid types.UID, status versionedPodStatus) bool {
 	return !ok || latest < status.version
 }
 
+func (m *manager) couldBeDeleted(uid types.UID, status v1.PodStatus) bool {
+	// The pod could be a static pod, so we should translate first.
+	pod, ok := m.podManager.GetPodByUID(uid)
+	if !ok {
+		return false
+	}
+	return !kubepod.IsMirrorPod(pod) && m.podDeletionSafety.OkToDeletePod(pod)
```
Comment (vishh) [inline pkg/kubelet/status/status_manager.go]: https://github.com/kubernetes/kubernetes/issues/41507 might be relevant here.

**Review by vishh — APPROVED** [review]

**Comment by vishh** [comment]
Here we go again :)

**Pull request merged**